Base bgp session parameters

cprabha opened the pull request:
Hi,

Please kindly review  base_bgp_session_parameters  for RT1.1 automation.

Thanks,
Prabha

cprabha commented:
Will open new pull request.

liulk commented:
Please revert the deletion of the *.testbed and *.binding files.

cprabha commented:
Hi Likai,

Please could you let me know how to revert testbed/binding files ?
I deleted in UI but now its now showing option to delete it.

Thanks,
Prabha



Juniper Business Use Only
From: Likai Liu ***@***.***>
Sent: Friday, May 27, 2022 2:39 PM
To: openconfig/featureprofiles ***@***.***>
Cc: Prabha Chethan ***@***.***>; State change ***@***.***>
Subject: Re: [openconfig/featureprofiles] Base bgp session parameters (PR #201)

[External Email. Be cautious of content]


Please revert the deletion of the *.testbed and *.binding files.

-
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/openconfig/featureprofiles/pull/201*issuecomment-1140058452__;Iw!!NEt6yMaO-gk!FOzo4Eefn2afMTudqHTe0KeO1dnraa6wqtjd-rTbHrlfWNIVixPU3Ko8PYSI5-dkYlO3qTgPeQXSNuC5ook8p3h1$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AHCEB5W2ELS6X2XQ5PKCUL3VME6HRANCNFSM5XFN7W6A__;!!NEt6yMaO-gk!FOzo4Eefn2afMTudqHTe0KeO1dnraa6wqtjd-rTbHrlfWNIVixPU3Ko8PYSI5-dkYlO3qTgPeQXSNuC5or4NokXz$>.
You are receiving this because you modified the open/close state.Message ID: ***@***.******@***.***>>

liulk commented:
I've only used the git command line and not the github UI. If you want to use UI, perhaps the GitHub Desktop App could help.
https://debug.to/1090/how-to-restore-deleted-file-in-github-website

This repo is also configured to do "squash and merge" for PRs, so I think just adding the files back should work. We can try this, and I'll take a look at the "files changed" in this PR and see what happens.

If a repo does not do "squash and merge," then I typically use `git rebase -i main` in the command line to produce a clean commit history, but you'd need to be careful with what you're doing, or you could accidentally lose a commit. The basic idea is to restore the file back from a past commit as a new commit, then reorder the commit history so the restoration commit either squash or fixup the commit that deleted it. The end result of the rebase should neutralize the unwanted deletion.

cprabha commented:
Hi Likai,

I have updated permissions on topology files. Please could you check.

Thanks,
Prabha



Juniper Business Use Only
From: Likai Liu ***@***.***>
Sent: Tuesday, May 31, 2022 8:24 AM
To: openconfig/featureprofiles ***@***.***>
Cc: Prabha Chethan ***@***.***>; State change ***@***.***>
Subject: Re: [openconfig/featureprofiles] Base bgp session parameters (PR #201)

[External Email. Be cautious of content]


It looks like the file permissions changed from 100644 → 100755 (from non-executable to executable) after you added the deleted files back. Please revert the file permissions as well. These files are not supposed to be executable.

-
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/openconfig/featureprofiles/pull/201*issuecomment-1142280988__;Iw!!NEt6yMaO-gk!FgK77IBCdlb-DgUlEp0d916fe3-6n7BWg6crhddnXmJigSvnOEvZpboXQ6HnZ1CZd874eFHFU97zXhS9vnnTm6c0$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AHCEB5WXFR6AKE6S7QZGP6TVMYVHZANCNFSM5XFN7W6A__;!!NEt6yMaO-gk!FgK77IBCdlb-DgUlEp0d916fe3-6n7BWg6crhddnXmJigSvnOEvZpboXQ6HnZ1CZd874eFHFU97zXhS9vlVJQLih$>.
You are receiving this because you modified the open/close state.Message ID: ***@***.******@***.***>>

cprabha commented:
Hi Likai,

Also on pull request, I could see couple of errors in static analysis.
Please could you share pointers to fix them.

Thanks,
Prabha





Juniper Business Use Only
From: Likai Liu ***@***.***>
Sent: Wednesday, June 1, 2022 9:59 AM
To: openconfig/featureprofiles ***@***.***>
Cc: Prabha Chethan ***@***.***>; State change ***@***.***>
Subject: Re: [openconfig/featureprofiles] Base bgp session parameters (PR #201)

[External Email. Be cautious of content]


Hi Prabha, after you changed the permissions in your local repo, please make sure to create a new commit with the permission change and push it to the PR.

-
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/openconfig/featureprofiles/pull/201*issuecomment-1143877987__;Iw!!NEt6yMaO-gk!BD-CTklaEDnRWmEW5JhbnbxtQoosgZ4ytGlg_t1ri_6qZOWl1ezS_7Ne-V8vysmb9Ul-PN416l9InlXJe2n5dXeZ$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AHCEB5R74XEITPGPRU6DTM3VM6JEZANCNFSM5XFN7W6A__;!!NEt6yMaO-gk!BD-CTklaEDnRWmEW5JhbnbxtQoosgZ4ytGlg_t1ri_6qZOWl1ezS_7Ne-V8vysmb9Ul-PN416l9InlXJe-r5Hb9L$>.
You are receiving this because you modified the open/close state.Message ID: ***@***.******@***.***>>

liulk commented:
I still haven't had a chance to review the code in detail, but I have another round of administrative comments:

 1. I still see the *.testbed and *.binding as executable. Please make sure you push the commit that sets them back to 0644?
 2. I don't think atedut_3 is needed for RT-1.1. Remove?
 3. For the test plan markdown, could you just convert the RT-1.1 from the shared Google doc using the exact wording instead of rewriting it? Here is an example from TE-1.1: https://github.com/openconfig/featureprofiles/blob/main/feature/interface/staticarp/ate_tests/static_arp_test/README.md
 4. Rename the test plan markdown to README.md?
 5. In the `*_test.go` file, it would help expedite the review if you could add the wording from the test plan as code comments.

Thanks for your patience.

cprabha commented:
Hi Likai,

1. I have updated permissions of topology files today and committed them
2. I have removed atedut_3.testbed file.
3.  I have updated README.md file as you have suggested.

I will update script with comments and commit it soon.

Thanks,
Prabha

liulk reviewed (changes requested):
Here is the first round of review addressing the more obvious issues. I'll follow up with a second round to look at the code in more detail. Again, thanks for the PR!

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,44 @@
+    Establish BGP session on ATE and DUT with table-based configuration to cover:
+        ○ Explicitly specified Router ID
+        ○ MD5 password
+        ○ Explicitly specified local and remote AS number
```
I remember in a prior commit, you had both eBGP and iBGP cases. Maybe we could add them back like this?

* Explicitly specified different local and remote AS number (eBGP)
* Explicitly specified same local and remote AS number (iBGP)

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,44 @@
+        ○ Explicitly specified retry interval
+    ● Validate session state and capabilities received on ATE.
+    ● Validate session state and capabilities received on DUT using telemetry.
+    ● Terminate BGP session using NOTIFICATION message and ensure that device telemetry correctly   reports the error codes.
```
Instead of a separate TODO below, just mark the TODO here like this:

* TODO: Terminate BGP session ...
  * Sending notification from ATE is not supported in the Ondatra framework. Waiting for Ixia support.

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,467 @@
+	}{
+		{
+			name:    "basic internal",
+			dutConf: bgpCreateNbr(dutAS, dutAS, 0, "", "", holdTime0, ConnTime0, NegHoldTime0),
```
Both bgpCreateNbr and configureATE have a long list of very similar arguments, and it's not apparent by reading the code here what the arguments are. It's possible someone might have swapped the order of some arguments without realizing it.

Consider creating a shared type like this:

```
type bgpParams struct {
  localAS int
  peerAS int
  nbrLocalAS int
  peerIP string
  authPwd string
  ...
}

func (params *bgpParams) forDUT() *telemetry.NetworkInstance_Protocol_Bgp {
  // This is the new bgpCreateNbr
}

func (params *bgpParams) forATE() *ondatra.ATETopology {
  // This is the new configureATE
}
```

Here in the test case, we can just have dutParams and ateParams, which we then convert to the OC struct and ATE topology in the body.

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,467 @@
+	t.Logf("Verify BGP session state : ESTABLISHED")
+	nbrPath.SessionState().Await(t, time.Second*100, telemetry.Bgp_Neighbor_SessionState_ESTABLISHED)
+
+	//TO DO: WAITING FOR THE FIX FROM ATE TO SEND NOTIFICATION TO DUT TO VERIFY ERROR CODE ON DUT
```
I like to use if false to comment out code, since text editors will still format the code.

```
// TODO: waiting for ATE to support notification.
if false {
  t.Log("Send notification from ATE")
  // Add code here to send notification.
  
  t.Log("Verify BGP session state: ACTIVE")
  ...
}
```

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,467 @@
+
+	//TO DO: WAITING FOR THE FIX FROM ATE TO SEND NOTIFICATION TO DUT TO VERIFY ERROR CODE ON DUT
+
+	/*t.Logf("Send Notification from ATE ")
```
Use `t.Log` if there is no formatting?

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,467 @@
+	NegHoldTime0 = 0
+)
+
+//Configure network instance
```
Note that for consistency's sake, we always have a space between `//` and the text.

cprabha commented:
Hi Likai,

I have updated README.md and base_bgp_session_parameters_test.go  files with all the review comments.
And pushed changes to github.

Looks like there is new check added now , and its failing due to "Executable Files".
I have not added any new testbed/binding files during this commit but its failing due to earlier commit of testbed/binding files.

Please could you let me know how to resolve this issue.

Thanks,
Prabha

SubhajitPalKeysight commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,417 @@
+	nbrPath.SessionState().Await(t, time.Second*100, telemetry.Bgp_Neighbor_SessionState_ESTABLISHED)
+
+	t.Log("Send Cease Notification from ATE to DUT")
+	// TODO: waiting for ATE to support notification.
```
ATE support of sending BGP notification has been provided. Please refer : https://github.com/openconfig/ondatra/pull/17

cprabha commented:
Thanks Subhajit. I have verified this and working fine.
-Prabha.



Juniper Business Use Only
From: Subhajit Pal ***@***.***>
Sent: Thursday, June 23, 2022 11:23 PM
To: openconfig/featureprofiles ***@***.***>
Cc: Prabha Chethan ***@***.***>; State change ***@***.***>
Subject: Re: [openconfig/featureprofiles] Base bgp session parameters (PR #201)

[External Email. Be cautious of content]


@SubhajitPalKeysight commented on this pull request.

________________________________

In feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go<https://urldefense.com/v3/__https:/github.com/openconfig/featureprofiles/pull/201*discussion_r905752503__;Iw!!NEt6yMaO-gk!HwS1w-FrDLMEa9IJLApuBQ1uM8DDs31_W9--ZUW2MkVvublWuwTXB8FmdFDk2MmAHehcQRPjPcxK47Mn2ncDmQtn$>:

> +     port1 := ate.Port(t, "port1")

+       topo := ate.Topology().New()

+       iDut1 := topo.AddInterface(ateAttrs.Name).WithPort(port1)

+ iDut1.IPv4().WithAddress(ateAttrs.IPv4CIDR()).WithDefaultGateway(dutAttrs.IPv4)

+       bgpDut1 := iDut1.BGP()

+ bgpDut1.AddPeer().WithPeerAddress(dutAttrs.IPv4).WithLocalASN(ateAS).WithTypeExternal()

+

+       t.Log("Pushing config to ATE and starting protocols...")

+       topo.Push(t)

+       topo.StartProtocols(t)

+

+       t.Log("Verify BGP session state : ESTABLISHED")

+       nbrPath.SessionState().Await(t, time.Second*100, telemetry.Bgp_Neighbor_SessionState_ESTABLISHED)

+

+       t.Log("Send Cease Notification from ATE to DUT")

+       // TODO: waiting for ATE to support notification.

ATE support of sending BGP notification has been provided. Please refer : openconfig/ondatra#17<https://urldefense.com/v3/__https:/github.com/openconfig/ondatra/pull/17__;!!NEt6yMaO-gk!HwS1w-FrDLMEa9IJLApuBQ1uM8DDs31_W9--ZUW2MkVvublWuwTXB8FmdFDk2MmAHehcQRPjPcxK47Mn2rXb7RkW$>

-
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/openconfig/featureprofiles/pull/201*pullrequestreview-1018031576__;Iw!!NEt6yMaO-gk!HwS1w-FrDLMEa9IJLApuBQ1uM8DDs31_W9--ZUW2MkVvublWuwTXB8FmdFDk2MmAHehcQRPjPcxK47Mn2rq56hH2$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AHCEB5TBLIOZWA7OPMKYBRTVQVH6BANCNFSM5XFN7W6A__;!!NEt6yMaO-gk!HwS1w-FrDLMEa9IJLApuBQ1uM8DDs31_W9--ZUW2MkVvublWuwTXB8FmdFDk2MmAHehcQRPjPcxK47Mn2izBASw1$>.
You are receiving this because you modified the open/close state.Message ID: ***@***.******@***.***>>

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,97 @@
+
+## Config Parameter Coverage
+
+    For prefix:
```
Format the "for prefix ... parameters" as shown in this example?

https://github.com/openconfig/featureprofiles/blob/main/feature/experimental/isis/ate_tests/base_adjacencies_test/README.md

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,97 @@
+        timers/config/keepalive-interval
+        timers/config/minimum-route-advertisement-interval
+
+## Telemetry Parameter coverage
```
Please ensure there is a blank line around the headers.

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,97 @@
+        transport/state/remote-port
+
+## Protocol/RPC Parameter coverage
+    BGP
```
Ditto for blank line and bullet point formatting.

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,97 @@
+
+## Notes
+
+   Explicit Keep alive timer in not supported. We do calculate relative keepalive timer from holdtimer value. Keepalive will be 1/3rd of negotiated hold timer.
```
Use regular paragraph?

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,425 @@
+// Copyright 2021 Google Inc.
```
Remove the executable bit of this file?

cprabha commented:
/gcbrun

xw-g commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,425 @@
+		RouterId: ygot.String(dutAttrs.IPv4),
+	}
+	dutConfNIPath := dut.Config().NetworkInstance(*deviations.DefaultNetworkInstance)
+	dutConfNIPath.Replace(t, ni)
```
Just reading this part, seems like the Replace() wipes out everything under existing default routing instance, right? If so, I think @liulk has a valid concern. If there are static route in the default routing instance, e.g., for management access, this Replace() probably will cut the device from connection.

cprabha replied:
I have updated script to configure only router id  and  NI instance type. please check

cprabha commented:
/gcbrun

cprabha commented:
Hi Likai,

Please could you let me know if any more changes needed for this pull request.

Thanks,
Prabha

cprabha commented:
Hi Likai,

Please could you let me know if any other changes needed for this PR.

Thanks,
Prabha

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,79 @@
+    BGP session should be established when different session parameters
+    are provided and session parameters are applied correspondingly.
+
+    *   Explicitly specified Router ID under bgp global level.
```
Since it's been a while since I looked at this test plan revision, I now have some additional thoughts.

It's now apparent to me that most of the parameters are disjoint, we can probably simplify the cases into just three or four major ones:

* iBGP - global AS
* eBGP - global AS
* eBGP - neighbor AS (does global AS still need to be specified?)
* (do we need an iBGP case for neighbor AS?)

The other parameters such as router ID, MD5 auth, holdtime and retry interval can just be set in all those three cases.

Also, let's state in the test plan which local and remote AS numbers have been picked for these cases.

cprabha replied:
@liulk Sure, I will work on these changes and submit changes.

cprabha replied:
@liulk I have summarized below test steps to automate  for RT1.1. Please could you confirm if these test steps are good to continue.

1.) iBGP -  
              i.) Verify with Global AS, 
              ii.) Verify with neighbor AS 
              iii.) Verify Session disconnect , receipt of cease error code
              iv.) Verify all session parameters - Route ID, MD5 Auth, Hold timer and Conn Retry Timer

2.) eBGP - 
              i.) Verify with Global AS, 
              ii.) Verify with neighbor AS 
              iii.) Verify Session disconnect , receipt of cease error code
              iv.) Verify all session parameters - Route ID, MD5 Auth, Hold timer and Conn Retry Timer

Thanks,
Prabha

liulk replied:
Hi Prabha, I updated README.md to clarify the procedure (you might need to rebase). Please take a look and let me know if that is good to continue.

cprabha replied:
@liulk , thank you, I am back from PTO.
I have checked README file, will update script 8/26 and submit it for review.

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,420 @@
+			tc.ateConf.StartProtocols(t)
+			t.Log("Verify BGP session state : ESTABLISHED")
+			nbrPath.SessionState().Await(t, time.Second*150, telemetry.Bgp_Neighbor_SessionState_ESTABLISHED)
+			stateDut := statePath.Get(t)
+			wantState := tc.wantState
+			if wantState == nil {
+				// if you don't specify a wantState, we just validate that the state corresponding to each config value is what it should be.
+				wantState = tc.dutConf
+			}
+			confirm.State(t, wantState, stateDut)
```
Just this?

```
confirm.State(t, tc.dutConf, statePath.Get(t))
```

cprabha replied:
If wantState == nil, we are assigning it to tc.dutConf.  Otherwise it checks for wantState defined in testcase.
please check below and let me know if this one is good.

	wantState := tc.wantState
			if wantState == nil {
				// if you don't specify a wantState, we just validate that the state corresponding to each config value is what it should be.
				wantState = tc.dutConf
			}
			confirm.State(t, wantState, stateDut)

cprabha replied:
updated these review comments, please check

liulk replied:
What I don't understand is that wantState looks largely the same as dutConf except it's missing some fields. But is that really necessary? Why can't we just use dutConf to confirm the state?

xw-g commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,79 @@
+
+*   DUT port-1 ------- ATE port-1    
+
+## Procedure
```
+1. LGTM. Thanks

dplore commented:
@liulk FYI,  @cprabha is out until Aug 23, will resume review then

cprabha commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:
```diff
@@ -0,0 +1,79 @@
+
+## Procedure
+
+*   Establish eBGP session between DUT and ATE 
```
You have modified README file , hence closing these comments.

cprabha commented:
@liulk , I have updated base bgp session parameters script and submitted changes. Please check.
And did not update anything on README file as you have already updated it.

Thanks,
Prabha

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,379 @@
+		{
+			name: "Test the eBGP session establishment: Global AS ",
+			dutConf: bgpCreateNbr(&bgpTestParams{localAS: dutAS, peerAS: ateAS,
+				routerID: dutAttrs.IPv4, holdTime: 100, connRetTime: 100, authPwd: "AUTHPASSWORD"}),
```
If the other parameters like routerID, holdTime, connRetTime and authPwd are all the same now, just inline them into bgpCreateNbr and omit them from bgpTestParams?

liulk commented:
Thanks for the contribution. I think we're very close now, just two more comments.

cprabha commented:
> 

Thanks Likai. I have updated both review comments provided , please could you check them.

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -49,11 +49,15 @@ var (
+	peerGrpName  = "BGP-PEER-GROUP"
+	authPassword = "AUTHPASSWORD"
+	holdTimer    = 100
+	connRettimer = 100
```
connRettimer -> connRetryTime?
ateHoldtimer -> ateHoldTime?

cprabha replied:
updated , please check

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -0,0 +1,358 @@
+}
+
+// Function to configure ATE configs based on args and returns ate topology handle.
+func configureATE(t *testing.T, ateParams *bgpTestParams) *ondatra.ATETopology {
```
Sorry just realized this, since connType is only used by configureATE, move it from bgpTestParams to an argument of configureATE?

Also, rather than a string, let's use const like this:

```
type connType string

const (
  connInternal connType = "INTERNAL"
  connExternal connType = "EXTERNAL"
)
```

cprabha replied:
updated, please check

liulk commented on feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:
```diff
@@ -206,12 +213,12 @@ func configureATE(t *testing.T, ateParams *bgpTestParams) *ondatra.ATETopology {
 	bgpDut1 := iDut1.BGP()
 
-	if ateParams.connType == "INTERNAL" {
+	if connectionType == "INTERNAL" {
```
```suggestion
	if connectionType == connInternal {
```

cprabha replied:
updated, thank you.

cprabha commented:
/gcbrun

cprabha commented:
/gcbrun

liulk commented:
I think the disco-idea-817 failure is because the go.mod/go.sum is outdated. I'm going to go ahead and merge.

liulk commented:
Thanks for the contribution!

cprabha commented:
Thank you.
-Prabha



Juniper Business Use Only
From: Likai Liu ***@***.***>
Sent: Wednesday, August 31, 2022 4:17 PM
To: openconfig/featureprofiles ***@***.***>
Cc: Prabha Chethan ***@***.***>; Mention ***@***.***>
Subject: Re: [openconfig/featureprofiles] Base bgp session parameters (PR #201)

[External Email. Be cautious of content]


I think the disco-idea-817 failure is because the go.mod/go.sum is outdated. I'm going to go ahead and merge.

-
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/openconfig/featureprofiles/pull/201*issuecomment-1233541472__;Iw!!NEt6yMaO-gk!HdghlTfxWXlpAqazgH1JadGeJdQaGiyAJNEqEtdYcccnxwWfYlnBFkl7s4t23aYz1KgmyqGaffaf-8UELPh5RcON$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AHCEB5Q2UJV6HOWGBOD77FLV37RYDANCNFSM5XFN7W6A__;!!NEt6yMaO-gk!HdghlTfxWXlpAqazgH1JadGeJdQaGiyAJNEqEtdYcccnxwWfYlnBFkl7s4t23aYz1KgmyqGaffaf-8UELEScpJnk$>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>
